File: Sources/WabeCore/Filter.swift
import simd

/// Chip→base axis maps, measured 2026-07-28 on Mac16,6 (gravity holds on three orientations +
/// two rotation-sign tests; see NOTES.md). The readout triad is LEFT-handed: chip +x = laptop
/// left, +y = front, +z = down. Base frame: X = right, Y = toward hinge, Z = up.
///
/// Accel (true vector) maps by full negation; gyro rates map by identity — a pseudo-vector picks
/// up the det(−I) = −1 under the improper transform, and both signs were confirmed empirically
/// (z: CCW spin, x: front-edge lift; y untested individually but pinned by the same global law).
@inlinable
public func accelChipToBase(_ v: SIMD3<Double>) -> SIMD3<Double> {
    -v
}

@inlinable
public func gyroChipToBase(_ v: SIMD3<Double>) -> SIMD3<Double> {
    v
}

/// Mahony-style complementary filter: gyro propagation, gravity correction on roll/pitch,
/// integral term absorbing gyro bias. Yaw is gyro-integrated only (unobservable without a
/// second reference) — consumers get it as *relative* heading via `recenter()`.
public final class MahonyFilter {
    public private(set) var q = simd_quatd(ix: 0, iy: 0, iz: 0, r: 1)  // base -> world
    public private(set) var gyroBias = SIMD3<Double>()
    public private(set) var stationary = false

    private let kp: Double
    private let ki: Double
    private var lastT: Double?
    private var headingRef = simd_quatd(ix: 0, iy: 0, iz: 0, r: 1)

    // Stationary detector state: windowed extremes over the last second.
    private var windowStart = 0.0
    private var accMagMin = Double.infinity, accMagMax = -Double.infinity
    private var gyroMax = 0.0
    private var stillSince: Double?
    private var gyroWindowSum = SIMD3<Double>()
    private var gyroWindowN = 0

    public init(kp: Double = 2.0, ki: Double = 0.05) {
        self.kp = kp
        self.ki = ki
    }

    public func update(accel: SIMD3<Double>, gyroDegS: SIMD3<Double>, t: Double) {
        guard let t0 = lastT else {
            lastT = t
            // Initialize attitude from gravity alone (yaw arbitrary = 0).
            q = Self.attitudeFromGravity(accel)
            headingRef = q
            return
        }
        let dt = t - t0
        lastT = t
        guard dt > 0, dt < 0.5 else { return }

        updateStationary(accel: accel, gyroDegS: gyroDegS, t: t)

        let omega = (gyroDegS - gyroBias) * (.pi / 180)

        // Gravity correction: error = measured down (body) x predicted down (body).
        var e = SIMD3<Double>()
        let accMag = length(accel)
        if accMag > 0.5, accMag < 1.5 {  // distrust accel during real shoving
            let measuredDown = -accel / accMag  // accel at rest reads +1 g opposite gravity
            let predictedDown = q.inverse.act(SIMD3(0, 0, -1))
            e = cross(measuredDown, predictedDown)
        }

        gyroBias -= e * ki * dt * (180 / .pi)
        let corrected = omega + e * kp
        let dq = simd_quatd(real: 0, imag: corrected) * q
        q = simd_normalize(simd_quatd(
            ix: q.imag.x + 0.5 * dq.imag.x * dt,
            iy: q.imag.y + 0.5 * dq.imag.y * dt,
            iz: q.imag.z + 0.5 * dq.imag.z * dt,
            r: q.real + 0.5 * dq.real * dt))
    }

    /// Laptop-intuitive angles, radians. Roll/pitch absolute (gravity), yaw relative to the last
    /// recenter. Signs: pitch + = front edge up, roll + = right side down, yaw + = CCW from above.
    public var rollPitchYaw: SIMD3<Double> {
        let rel = headingRef.inverse * q
        let R = simd_matrix3x3(q)
        let aboutX = atan2(R[1][2], R[2][2])   // math-euler angle about base X (left-right axis)
        let aboutY = -asin(max(-1, min(1, R[0][2])))  // about base Y (front-back axis)
        let Rrel = simd_matrix3x3(rel)
        let yaw = atan2(Rrel[0][1], Rrel[0][0])
        return SIMD3(aboutY, -aboutX, yaw)  // (roll, pitch, yaw) in laptop terms
    }

    public func recenter() { headingRef = q }

    private static func attitudeFromGravity(_ accel: SIMD3<Double>) -> simd_quatd {
        let up = normalize(accel)  // at rest accel points opposite gravity = world up
        return simd_quatd(from: up, to: SIMD3(0, 0, 1))
    }

    private func updateStationary(accel: SIMD3<Double>, gyroDegS: SIMD3<Double>, t: Double) {
        let m = length(accel)
        accMagMin = min(accMagMin, m)
        accMagMax = max(accMagMax, m)
        gyroMax = max(gyroMax, length(gyroDegS - gyroBias))
        gyroWindowSum += gyroDegS  // raw, pre-bias: the window mean of a still window IS the bias
        gyroWindowN += 1
        guard t - windowStart >= 1.0 else { return }

        let still = (accMagMax - accMagMin) < 0.02 && gyroMax < 1.5
        if still {
            if stillSince == nil { stillSince = t }
            // Stationary bias refresh. This is the only correction path for z (yaw) bias —
            // gravity error has no z-component — and it sharpens x/y beyond what ki achieves.
            // Blend rather than snap so a borderline-still window can't kick the rates.
            if gyroWindowN > 0 {
                let mean = gyroWindowSum / Double(gyroWindowN)
                gyroBias += (mean - gyroBias) * 0.5
            }
        } else {
            stillSince = nil
        }
        stationary = still && (t - (stillSince ?? t)) >= 0.0

        windowStart = t
        accMagMin = .infinity
        accMagMax = -.infinity
        gyroMax = 0
        gyroWindowSum = SIMD3<Double>()
        gyroWindowN = 0
    }
}

/// Screen-plane pose: base attitude ⊕ lid angle. The hinge axis is +x of the base frame
/// (PROVISIONAL, same caveat as above). Lid angle 0 = closed, ~110 = typical open. The screen
/// normal at lid angle θ points away from the display face.
public enum ScreenPose {
    public static func screenNormal(baseQ: simd_quatd, lidDeg: Double) -> SIMD3<Double> {
        let theta = (180 - lidDeg) * .pi / 180  // closed lid: normal flat against keyboard
        let hinge = simd_quatd(angle: theta, axis: SIMD3(1, 0, 0))
        // Screen face normal when fully flat-open (180°) would be straight up in base frame.
        let normalInBase = hinge.act(SIMD3(0, 0, 1))
        return baseQ.act(normalInBase)
    }
}

File: Sources/libwabe/filter.c
// Filter core: chip->base axis maps, timestamp merge, VQF (via the C shim over the vendored
// C++), heading reference, and pose extraction. Pure computation — no I/O.
#include "include/wabe.h"
#include "vendor/wabe_vqf.h"

#include <math.h>
#include <stdlib.h>
#include <string.h>

#define DEG2RAD (M_PI / 180.0)
#define RAD2DEG (180.0 / M_PI)
#define G_TO_MS2 9.80665

struct wabe_filter {
    wvqf *vqf;
    double q[4];    // base -> world, w x y z
    double ref[4];  // heading reference (recenter)
    double bias[3]; // rad/s
    int rest;
    double last_accel[3]; // base frame, g (zero-order hold)
    int have_accel;
    double lid_deg;
    int first;
};

// Chip->base axis maps, measured on Mac16,6 (see NOTES.md): the readout triad is left-handed
// (chip +x = laptop left, +y = front, +z = down). Accel (true vector) maps by full negation;
// gyro (pseudo-vector) maps by identity, so gyro samples are used as-is below.
static void accel_chip_to_base(const double v[3], double out[3])
{
    out[0] = -v[0];
    out[1] = -v[1];
    out[2] = -v[2];
}

// q multiply, (w,x,y,z)
static void quat_mul(const double a[4], const double b[4], double out[4])
{
    out[0] = a[0] * b[0] - a[1] * b[1] - a[2] * b[2] - a[3] * b[3];
    out[1] = a[0] * b[1] + a[1] * b[0] + a[2] * b[3] - a[3] * b[2];
    out[2] = a[0] * b[2] - a[1] * b[3] + a[2] * b[0] + a[3] * b[1];
    out[3] = a[0] * b[3] + a[1] * b[2] - a[2] * b[1] + a[3] * b[0];
}

static void quat_conj(const double q[4], double out[4])
{
    out[0] = q[0];
    out[1] = -q[1];
    out[2] = -q[2];
    out[3] = -q[3];
}

// Row-major rotation matrix of a unit quaternion (base->world).
static void quat_to_mat(const double q[4], double m[3][3])
{
    const double w = q[0], x = q[1], y = q[2], z = q[3];
    m[0][0] = 1 - 2 * (y * y + z * z);
    m[0][1] = 2 * (x * y - w * z);
    m[0][2] = 2 * (x * z + w * y);
    m[1][0] = 2 * (x * y + w * z);
    m[1][1] = 1 - 2 * (x * x + z * z);
    m[1][2] = 2 * (y * z - w * x);
    m[2][0] = 2 * (x * z - w * y);
    m[2][1] = 2 * (y * z + w * x);
    m[2][2] = 1 - 2 * (x * x + y * y);
}

wabe_filter *wabe_filter_new(double sample_hz)
{
    wabe_filter *f = calloc(1, sizeof(*f));
    if (!f)
        return NULL;
    f->vqf = wvqf_create(1.0 / (sample_hz > 0 ? sample_hz : 795.0));
    f->q[0] = 1;
    f->ref[0] = 1;
    f->lid_deg = -1;
    f->first = 1;
    return f;
}

void wabe_filter_free(wabe_filter *f)
{
    if (!f)
        return;
    wvqf_destroy(f->vqf);
    free(f);
}

static void filter_update(wabe_filter *f, const double gyro_chip[3])
{
    double gyr[3], acc[3];
    for (int i = 0; i < 3; i++) {
        gyr[i] = gyro_chip[i] * DEG2RAD;  // gyro base frame == chip frame
        acc[i] = f->last_accel[i] * G_TO_MS2;
    }
    wvqf_update(f->vqf, gyr, acc);
    wvqf_quat6d(f->vqf, f->q);
    wvqf_bias(f->vqf, f->bias);
    f->rest = wvqf_rest(f->vqf);
    if (f->first) {
        f->first = 0;
        memcpy(f->ref, f->q, sizeof(f->ref));
    }
}

void wabe_filter_feed(wabe_filter *f, const wabe_sample *accel, size_t na,
                      const wabe_sample *gyro, size_t ng)
{
    size_t ai = 0;
    for (size_t gi = 0; gi < ng; gi++) {
        while (ai < na && accel[ai].t <= gyro[gi].t) {
            accel_chip_to_base(accel[ai].v, f->last_accel);
            f->have_accel = 1;
            ai++;
        }
        if (f->have_accel)
            filter_update(f, gyro[gi].v);
    }
    while (ai < na) {
        accel_chip_to_base(accel[ai].v, f->last_accel);
        f->have_accel = 1;
        ai++;
    }
}

void wabe_filter_set_lid(wabe_filter *f, double deg) { f->lid_deg = deg; }

void wabe_filter_recenter(wabe_filter *f) { memcpy(f->ref, f->q, sizeof(f->ref)); }

void wabe_filter_pose(const wabe_filter *f, double t, wabe_pose *out)
{
    memset(out, 0, sizeof(*out));
    out->t = t;
    memcpy(out->q, f->q, sizeof(out->q));
    for (int i = 0; i < 3; i++)
        out->bias[i] = f->bias[i] * RAD2DEG;
    out->stationary = f->rest;
    out->lid_deg = f->lid_deg;

    // Laptop-intuitive angles. Roll/pitch absolute (gravity), yaw relative to the last
    // recenter. Signs: pitch + = front edge up, roll + = right side down, yaw + = CCW from
    // above.
    double m[3][3];
    quat_to_mat(f->q, m);
    const double about_x = atan2(m[2][1], m[2][2]); // about base X (left-right axis)
    const double about_y = -asin(fmax(-1.0, fmin(1.0, m[2][0]))); // about base Y (front-back)

    double ref_inv[4], rel[4], mrel[3][3];
    quat_conj(f->ref, ref_inv);
    quat_mul(ref_inv, f->q, rel);
    quat_to_mat(rel, mrel);
    const double yaw = atan2(mrel[1][0], mrel[0][0]);

    out->rpy[0] = about_y * RAD2DEG;
    out->rpy[1] = -about_x * RAD2DEG;
    out->rpy[2] = yaw * RAD2DEG;

    // Screen normal: base attitude ⊕ lid angle. Hinge axis = base +X; at lid angle L the
    // face normal in base frame is (0,0,1) rotated about X by (180-L) degrees.
    if (f->lid_deg >= 0) {
        const double theta = (180.0 - f->lid_deg) * DEG2RAD;
        const double nb[3] = {0, -sin(theta), cos(theta)};
        out->n[0] = m[0][0] * nb[0] + m[0][1] * nb[1] + m[0][2] * nb[2];
        out->n[1] = m[1][0] * nb[0] + m[1][1] * nb[1] + m[1][2] * nb[2];
        out->n[2] = m[2][0] * nb[0] + m[2][1] * nb[1] + m[2][2] * nb[2];
    }
}
